docs/api/standard_library_types.md

```diff
@@ -922,9 +928,125 @@ except ValidationError as e:
+    dessert: Annotated[
+        Union[
+            Annotated[ApplePie, Tag('apple')],
+            Annotated[PumpkinPie, Tag('pumpkin')],
+        ],
+        CallableDiscriminator(get_discriminator_value),
+    ]
+
+
+apple_variation = ThanksgivingDinner.model_validate(
+    {'dessert': {'fruit': 'apple', 'time_to_cook': 60, 'num_ingredients': 8}}
+)
+print(repr(apple_variation))
+"""
+ThanksgivingDinner(dessert=ApplePie(time_to_cook=60, num_ingredients=8, fruit='apple'))
+"""
+
+pumpkin_variation = ThanksgivingDinner.model_validate(
+    {
+        'dessert': {
+            'filling': 'pumpkin',
+            'time_to_cook': 40,
+            'num_ingredients': 6,
+        }
+    }
+)
+print(repr(pumpkin_variation))
+"""
+ThanksgivingDinner(dessert=PumpkinPie(time_to_cook=40, num_ingredients=6, filling='pumpkin'))
+"""
+```
+
+`CallableDiscriminators` can also be used to validate `Union` types with combinations of models and primitive types.
+
+For example:
+
+```py requires="3.8"
+from typing import Any, Union
+
+from typing_extensions import Annotated
+
+from pydantic import BaseModel, CallableDiscriminator, Tag
+
+
+def model_x_discriminator(v: Any) -> str:
+    if isinstance(v, str):
+        return 'str'
+    if isinstance(v, (dict, BaseModel)):
+        return 'model'
+
+
+class DiscriminatedModel(BaseModel):
+    x: Annotated[
+        Union[
+            Annotated[str, Tag('str')],
+            Annotated['DiscriminatedModel', Tag('model')],
+        ],
+        CallableDiscriminator(
+            model_x_discriminator,
+            custom_error_type='invalid_union_member',
+            custom_error_message='Invalid union member',
+            custom_error_context={'discriminator': 'str_or_model'},
+        ),
+    ]
+
+
+data = {'x': {'x': {'x': 'a'}}}
+m = DiscriminatedModel.model_validate(data)
+assert m == (
+    DiscriminatedModel(x=DiscriminatedModel(x=DiscriminatedModel(x='a')))
+)
+assert m.model_dump() == data
+```
+
 !!! note
     Using the [`typing.Annotated` fields syntax](../concepts/json_schema.md#typingannotated-fields) can be handy to regroup
-    the `Union` and `discriminator` information. See below for an example!
+    the `Union` and `discriminator` information. See the next example for more details.
+
+    There are a few ways to set a discriminator for a field, all varying slightly in syntax.
+
+    For `str` discriminators:
+        * `some_field: Union[...] = Field(discriminator='my_discriminator'`
+        * `some_field: Annotated[Union[...], Field(discriminator='my_discriminator')]`
+    For `CallableDiscriminator` discriminators:
+        * `some_field: Union[...] = Field(discriminator=CallableDiscriminator(...))`
+        * `some_field: Annotated[Union[...], CallableDiscriminator(...)]`
+        * `some_field: Annotated[Union[...], Field(discriminator=CallableDiscriminator(...))]`
 
 !!! warning
     Discriminated unions cannot be used with only a single variant, such as `Union[Cat]`.
```

Comment:
Not sure if this applies to callable discriminators. And there's a non-contrived (imo, anyway) use case for this — faster failing if a specific key/value pair is missing. I mean, it will rarely be relevant, but you could imagine an API where you have many object types but you only want to allow one variant in a certain position, and don't need to validate/produce a complex error message if the tag is just wrong.

I don't think we need to necessarily document that usage pattern, but I'm just thinking it maybe makes sense to update this warning given that; I defer to you @sydney-runkle. (I also don't promise it actually works with just a single case, I just think it does based on the implementation 😄)

Reply:
@dmontagu,

I think the warning still holds:

```py
from typing import Any, Literal, Union

from typing_extensions import Annotated

from pydantic import BaseModel, TypeAdapter, CallableDiscriminator, Tag


def get_discriminator_value(v: Any) -> str:
    if isinstance(v, dict):
        return v.get('pet_type')
    return getattr(v, 'fruit', None)


class Cat(BaseModel):
    pet_type: Literal['cat'] = 'cat'


ta = TypeAdapter(Annotated[Union[Annotated[Cat, Tag('cat')]], CallableDiscriminator(get_discriminator_value)])
#> CallableDiscriminator must be used with a Union type, not <class '__main__.Cat'>
```

Reply:
Resolving + leaving the warning in place 😢